client.go

```diff
@@ -0,0 +1,260 @@
+	// Send() is called, you would specify 250 here.
+	SampleRate uint
+
+	// APIHost is the hostname for the Honeycomb API server to which to send this
+	// event. default: https://api.honeycomb.io/
+	APIHost string
+
+	// Transmission allows you to override what happens to events after you call
+	// Send() on them. By default, events are asynchronously sent to the
+	// Honeycomb API. You can use the MockOutput included in this package in
+	// unit tests, or use the transmission.WriterSender to write events to
+	// STDOUT or to a file when developing locally.
+	Transmission transmission.Sender
+
+	// Logger defaults to nil and the SDK is silent. If you supply a logger here
+	// (or set it to &DefaultLogger{}), some debugging output will be emitted.
+	// Intended for human consumption during development to understand what the
+	// SDK is doing and diagnose trouble emitting events.
+	Logger Logger
+}
+
+// NewClient creates a Client with defaults correctly set
+func NewClient(conf ClientConfig) (*Client, error) {
+	if conf.SampleRate == 0 {
+		conf.SampleRate = defaultSampleRate
+	}
+	if conf.APIHost == "" {
+		conf.APIHost = defaultAPIHost
+	}
+	if conf.Dataset == "" {
+		conf.Dataset = defaultDataset
+	}
+
+	c := &Client{
+		logger: conf.Logger,
+	}
+	c.ensureLogger()
+
+	if conf.Transmission == nil {
+		c.transmission = &transmission.Honeycomb{
+			MaxBatchSize:         DefaultMaxBatchSize,
+			BatchTimeout:         DefaultBatchTimeout,
+			MaxConcurrentBatches: DefaultMaxConcurrentBatches,
+			PendingWorkCapacity:  DefaultPendingWorkCapacity,
+			UserAgentAddition:    UserAgentAddition,
+			Logger:               c.logger,
+			Metrics:              sd,
+		}
+	} else {
+		c.transmission = conf.Transmission
+	}
+	if err := c.transmission.Start(); err != nil {
+		c.logger.Printf("transmission client failed to start: %s", err.Error())
+		return nil, err
+	}
+
+	c.builder = &Builder{
+		WriteKey:   conf.APIKey,
+		Dataset:    conf.Dataset,
+		SampleRate: conf.SampleRate,
+		APIHost:    conf.APIHost,
+		dynFields:  make([]dynamicField, 0, 0),
+		fieldHolder: fieldHolder{
+			data: make(map[string]interface{}),
+		},
+		client: c,
+	}
+
+	return c, nil
+}
+
+func (c *Client) ensureTransmission() {
+	c.oneTx.Do(func() {
+		if c.transmission == nil {
+			c.transmission = &transmission.DiscardSender{}
+			c.transmission.Start()
+		}
+	})
+}
+
+func (c *Client) ensureLogger() {
+	c.oneLogger.Do(func() {
+		if c.logger == nil {
+			c.logger = &nullLogger{}
+		}
+	})
+}
+
+func (c *Client) ensureBuilder() {
+	c.oneBuilder.Do(func() {
+		if c.builder == nil {
+			c.builder = &Builder{
+				dynFields: make([]dynamicField, 0, 0),
+				fieldHolder: fieldHolder{
+					data: make(map[string]interface{}),
+				},
+				client: c,
+			}
+		}
+	})
+}
+
+// Close waits for all in-flight messages to be sent. You should
+// call Close() before app termination.
+func (c *Client) Close() {
+	if c == nil {
```

Comment:
except that by function of an empty Client{} being a working (discard) client, creating the thing then throwing it out means that a nil `c.NewBulider().NewEvent()` would still create a valid event. 
but ...
I am also on the fence and am comfortable saying that this should panic. It's a tall order - IMO an instrumentation library should _never_ panic - blowing up the service because the instrumentation is broken feels extraordinarily rude. But I also agree that `c` being `nil` is 99% likely an error (and one that should be found, because the absence of instrumentation may incorrectly be interpreted as the instrumented thing having not run), so we should help find the error via panic.

I'm going to remove the nil receiver checks.